src/Bicep.Core/Emit/TemplateWriter.cs

```diff
@@ -473,10 +474,20 @@ private void EmitDependsOn(JsonTextWriter memoryWriter, DeclaredSymbol declaredS
                 return;
             }
 
+            // De-duplicate dependencies which are based on a resource collection
+            var dedupedDependencies = new HashSet<ResourceDependency>();
+            foreach (var dependency in dependencies)
+            {
+                if (!dedupedDependencies.Any(d => d.Resource == dependency.Resource && (d.IndexExpression as IntegerLiteralSyntax)?.Value == (dependency.IndexExpression as IntegerLiteralSyntax)?.Value))
```

Comment:
I think it'll be better to write a comparator  and use Linq Distinct method.

Reply:
good point! I changed that. Although I can't say if this is the right approach as I don't really understand what type other than IntegerLiteralSyntax the IndexExpression could be

Reply:
The problem is that the index expression can be any expression type. You could use any operator, function call, etc. The number of combinations is pretty much infinite (even if in most cases, the tree will be fairly simple).

As @miqm was suggesting, the best approach is to implement a generic `SyntaxBase` comparer. The comparer would implement deep equals and deep hash code that should work on all possible valid and invalid trees. On top of that I would want to avoid introducing a separate syntax recursion mechanism and reuse the knowledge of the tree structure that is represented in `SyntaxVisitor` but without introducing too many memory allocations in the implementation.

Reply:
Yeah I figured that there is more complexity to this. I'll close this PR